### PR TITLE
fix: Suppress the warning from boost to fix build failure.

### LINKED ISSUE #4771: Fail to build tiflash due to unused variables
### 1. Minimal reproduce step (Required)

clone the repo and try to build it.
```
mkdir build && cd build
CC=clang CXX=clang++ cmake -G Ninja -DCMAKE_BUILD_TYPE=Debug -DCMAKE_EXPORT_COMPILE_COMMANDS=YES ../
```

### 2. What did you expect to see? (Required)
Success
### 3. What did you see instead (Required)
```
[0/2] Re-checking globbed directories...
[2/268] Building CXX object dbms/src/Functions/CMakeFiles/clickhouse_functions.dir/FunctionsGeo.cpp.o
FAILED: dbms/src/Functions/CMakeFiles/clickhouse_functions.dir/FunctionsGeo.cpp.o 
/usr/bin/ccache /home/jun/opt/clang/bin/clang++ -DBOOST_CONTEXT_EXPORT="" -DBOOST_CONTEXT_NO_LIB -DBOOST_CONTEXT_STATIC_LINK -DBOOST_FIBER_NO_LIB -DBOOST_FIBER_STATIC_LINK -DBOOST_SYSTEM_NO_DEPRECATED -DCARES_STATICLIB -DFIU_ENABLE -DLZ4_DISABLE_DEPRECATE_WARNINGS=1 -DPOCO_STATIC -DPOCO_UNBUNDLED_ZLIB -DSTACK_LINE_READER_BUFFER_SIZE=1024 -DTIFLASH_COMPILER_VPCLMULQDQ_SUPPORT=1 -DTIFLASH_ENABLE_AVX512_SUPPORT=1 -DTIFLASH_ENABLE_AVX_SUPPORT=1 -DTIFLASH_SOURCE_PREFIX=\"/home/jun/dev/tiflash\" -I/home/jun/dev/tiflash/contrib/libdivide -I/home/jun/dev/tiflash/contrib/libmetrohash/src -I/home/jun/dev/tiflash/contrib/libfarmhash -I/home/jun/dev/tiflash/dbms/src -I/home/jun/dev/tiflash/contrib/tiflash-proxy/raftstore-proxy/ffi/src -I/home/jun/dev/tiflash/build/dbms/src -I/home/jun/dev/tiflash/contrib/double-conversion -I/home/jun/dev/tiflash/contrib/boost -I/home/jun/dev/tiflash/contrib/xxHash -I/home/jun/dev/tiflash/contrib/libpcg-random/include -I/home/jun/dev/tiflash/contrib/libcityhash/include -I/home/jun/dev/tiflash/libs/libcommon/include -I/home/jun/dev/tiflash/build/libs/libcommon/include -I/home/jun/dev/tiflash/libs/libpocoext/include -I/home/jun/dev/tiflash/contrib/poco/Data/include -I/home/jun/dev/tiflash/contrib/poco/Foundation/include -I/home/jun/dev/tiflash/contrib/zlib-ng -I/home/jun/dev/tiflash/build/contrib/zlib-ng -I/home/jun/dev/tiflash/contrib/poco/Util/include -I/home/jun/dev/tiflash/contrib/poco/XML/include -I/home/jun/dev/tiflash/contrib/poco/JSON/include -I/home/jun/dev/tiflash/contrib/poco/Net/include -I/home/jun/dev/tiflash/contrib/cctz/include -I/home/jun/dev/tiflash/contrib/jemalloc-cmake/include -I/home/jun/dev/tiflash/contrib/jemalloc-cmake/include_linux_x86_64 -I/home/jun/dev/tiflash/contrib/cpu_features/include -I/home/jun/dev/tiflash/contrib/libcpuid/include -I/home/jun/dev/tiflash/contrib/lz4/lib -I/home/jun/dev/tiflash/contrib/zstd/lib -I/home/jun/dev/tiflash/contrib/client-c/third_party/libfiu/libfiu -I/home/jun/dev/tiflash/contrib/prometheus-cpp/core/include -I/home/jun/dev/tiflash/build/contrib/prometheus-cpp-cmake/core/include -I/home/jun/dev/tiflash/contrib/prometheus-cpp/push/include -I/home/jun/dev/tiflash/build/contrib/prometheus-cpp-cmake/push/include -I/home/jun/dev/tiflash/contrib/prometheus-cpp/pull/include -I/home/jun/dev/tiflash/build/contrib/prometheus-cpp-cmake/pull/include -I/home/jun/dev/tiflash/contrib/cpptoml -I/home/jun/dev/tiflash/contrib/poco/NetSSL_OpenSSL/include -I/home/jun/dev/tiflash/contrib/poco/Crypto/include -I/home/jun/dev/tiflash/contrib/xxHash/cmake_unofficial/.. -I/home/jun/dev/tiflash/contrib/kvproto/cpp -I/home/jun/dev/tiflash/contrib/kvproto/cpp/kvproto -I/home/jun/dev/tiflash/contrib/grpc/include -I/home/jun/dev/tiflash/contrib/client-c/include -I/home/jun/dev/tiflash/build/contrib/grpc/third_party/cares/cares -I/home/jun/dev/tiflash/contrib/grpc/third_party/cares/cares -I/home/jun/dev/tiflash/contrib/tipb/cpp -I/home/jun/dev/tiflash/build/dbms/src/Storages/DeltaMerge/File/dtpb -I/home/jun/dev/tiflash/contrib/libbtrie/include -I/home/jun/dev/tiflash/contrib/abseil-cpp -I/home/jun/dev/tiflash/libs/libconsistent-hashing -I/home/jun/dev/tiflash/build/contrib/libfarmhash -isystem /home/jun/dev/tiflash/contrib/fmtlib-cmake/../fmtlib/include -isystem /home/jun/dev/tiflash/libs/libsymbolization/include -isystem /home/jun/dev/tiflash/contrib/boringssl/include -isystem /home/jun/dev/tiflash/contrib/protobuf/src -isystem /home/jun/dev/tiflash/contrib/re2 -isystem /home/jun/dev/tiflash/build/contrib/re2-cmake -D__GLIBCXX_BITSIZE_INT_N_0=128 -D__GLIBCXX_TYPE_INT_N_0=__int128  -D_GLIBCXX_USE_CXX11_ABI=1 -pipe -msse4.1 -msse4.2 -mpopcnt  -fno-omit-frame-pointer  -Wall -Wno-unused-command-line-argument  -Wnon-virtual-dtor  -Wextra -Werror -g -g3 -ggdb3 -O0 -fverbose-asm -fno-inline  -fPIC -DDUMMY_BACKTRACE -pthread -std=gnu++17 -MD -MT dbms/src/Functions/CMakeFiles/clickhouse_functions.dir/FunctionsGeo.cpp.o -MF dbms/src/Functions/CMakeFiles/clickhouse_functions.dir/FunctionsGeo.cpp.o.d -o dbms/src/Functions/CMakeFiles/clickhouse_functions.dir/FunctionsGeo.cpp.o -c /home/jun/dev/tiflash/dbms/src/Functions/FunctionsGeo.cpp
In file included from /home/jun/dev/tiflash/dbms/src/Functions/FunctionsGeo.cpp:22:
In file included from /home/jun/dev/tiflash/dbms/src/Functions/GeoUtils.h:38:
In file included from /home/jun/dev/tiflash/contrib/boost/boost/geometry.hpp:17:
In file included from /home/jun/dev/tiflash/contrib/boost/boost/geometry/geometry.hpp:58:
In file included from /home/jun/dev/tiflash/contrib/boost/boost/geometry/strategies/strategies.hpp:37:
In file included from /home/jun/dev/tiflash/contrib/boost/boost/geometry/strategies/intersection_strategies.hpp:19:
In file included from /home/jun/dev/tiflash/contrib/boost/boost/geometry/policies/relate/intersection_policy.hpp:21:
/home/jun/dev/tiflash/contrib/boost/boost/geometry/policies/relate/intersection_points.hpp:71:33: error: variable 'count_a' set but not used [-Werror,-Wunused-but-set-variable]
        unsigned int index = 0, count_a = 0, count_b = 0;
                                ^
/home/jun/dev/tiflash/contrib/boost/boost/geometry/policies/relate/intersection_points.hpp:71:46: error: variable 'count_b' set but not used [-Werror,-Wunused-but-set-variable]
        unsigned int index = 0, count_a = 0, count_b = 0;
                                             ^
2 errors generated.
[10/268] Building CXX object dbms/src/Functions/CMakeFiles/clickhouse_functions.dir/divide.cpp.o^C
ninja: build stopped: interrupted by user.
```

### 4. What is your TiFlash version? (Required)
main branch

Clang version:
```
clang version 15.0.0 (https://github.com/llvm/llvm-project/ 4ffd0b6fde4da9a3ba4ee3a189504ce84c118b1c)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/jun/opt/clang/bin
```

### PR DESCRIPTION
**Pull request by junaire**
Signed-off-by: Jun Zhang <jun@junz.org>

### What problem does this PR solve?

Issue Number: close #4771 

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Functions/FunctionsGeo.cpp [hunk 1]
```diff
@@ -12,6 +12,8 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#pragma GCC diagnostic ignored "-Wunused-but-set-variable"
```
Comment (JaySon-Huang) [inline dbms/src/Functions/FunctionsGeo.cpp:15]: I think it is better to make this suppression only apply to `GeoUtils.h` line 32~44. We wrap those suppressions with "diagnostic push/pop" to keep the TiFlash code clean for those warnings.

https://github.com/pingcap/tiflash/blob/c6740e5c8d184d854447cd7d8e04939efe99de75/dbms/src/Functions/GeoUtils.h#L32-L44

**Review by fuzhe1989 — APPROVED** [review]

**Review by JaySon-Huang — APPROVED** [review]

**Comment by jiaqizho** [comment]
@JaySon-Huang  @fuzhe1989 Can't build-in clang.

```
/dbms/src/Functions/GeoUtils.h:33:32: error: unknown warning group '-Wunused-but-set-variable', ignored [-Werror,-Wunknown-warning-option]
#pragma GCC diagnostic ignored "-Wunused-but-set-variable"
                               ^
```

**Comment by junaire** [comment]
> @JaySon-Huang  @fuzhe1989 Can't build-in clang.
> 
> ```
> /dbms/src/Functions/GeoUtils.h:33:32: error: unknown warning group '-Wunused-but-set-variable', ignored [-Werror,-Wunknown-warning-option]
> #pragma GCC diagnostic ignored "-Wunused-but-set-variable"
>                                ^
> ```

May I ask your clang version? For me everything works fine for Clang trunk.
In fact, this flag is recently added in https://reviews.llvm.org/D100581, and I was surprised they don't even provide a release note...

**Comment by JaySon-Huang** [comment]
> @JaySon-Huang @fuzhe1989 Can't build-in clang.
> 
> ```
> /dbms/src/Functions/GeoUtils.h:33:32: error: unknown warning group '-Wunused-but-set-variable', ignored [-Werror,-Wunknown-warning-option]
> #pragma GCC diagnostic ignored "-Wunused-but-set-variable"
>                                ^
> ```

@jiaqizho  What's your error message for compile error? The CI build TiFlash with Clang 13.0.0 and give no warnings/errors https://ci.pingcap.net/blue/organizations/jenkins/tiflash-build-common/detail/tiflash-build-common/4096/pipeline/150

**Comment by junaire** [comment]
I just realized that the minimal toolchain requirement were not matched in the CMakelists.txt and README. That said cmake will only complain when clang < 5 or gcc < 7. I guess that will confuse some folks especially who first work on the project. If we really just want guarantee tiflash compiles in recent compilers, maybe we should align the requirements in CMakelists.txt @JaySon-Huang @fuzhe1989 @ywqzzy

**Comment by JaySon-Huang** [comment]
Now we only declare the minimal toolchain requirement in the README.md (and collapsed by default). We have not changed the related warnings in CMakeLists.txt yet.

/cc @SchrodingerZhu 

image

**Comment by jiaqizho** [comment]
> > @JaySon-Huang @fuzhe1989 Can't build-in clang.
> > ```
> > /dbms/src/Functions/GeoUtils.h:33:32: error: unknown warning group '-Wunused-but-set-variable', ignored [-Werror,-Wunknown-warning-option]
> > #pragma GCC diagnostic ignored "-Wunused-but-set-variable"
> >                                ^
> > ```
> 
> @jiaqizho What's your error message for compile error? The CI build TiFlash with Clang 13.0.0 and give no warnings/errors https://ci.pingcap.net/blue/organizations/jenkins/tiflash-build-common/detail/tiflash-build-common/4096/pipeline/150

Apple clang version 12.0.5 (clang-1205.0.22.11)

**Comment by junaire** [comment]
> > > @JaySon-Huang @fuzhe1989 Can't build-in clang.
> > > ```
> > > /dbms/src/Functions/GeoUtils.h:33:32: error: unknown warning group '-Wunused-but-set-variable', ignored [-Werror,-Wunknown-warning-option]
> > > #pragma GCC diagnostic ignored "-Wunused-but-set-variable"
> > >                                ^
> > > ```
> > 
> > 
> > @jiaqizho What's your error message for compile error? The CI build TiFlash with Clang 13.0.0 and give no warnings/errors https://ci.pingcap.net/blue/organizations/jenkins/tiflash-build-common/detail/tiflash-build-common/4096/pipeline/150
> 
> Apple clang version 12.0.5 (clang-1205.0.22.11)

Good, that's the issue I guess. I don't have a mac so I can't really test it, but the following godbolt link shows that this flag is only supported after clang 13: https://godbolt.org/z/hWr5bss85 Maybe try to use homebrew to get yourself a more recent compiler?

**Comment by JaySon-Huang** [comment]
We can add a line `#pragma GCC diagnostic ignored "-Wunknown-warning-option"` before "-Wunused-but-set-variable" to make it compatible with older clang https://github.com/pingcap/tiflash/pull/4784/files

**Pull request merged**